design-proposals/orch-cli.md

```diff
@@ -0,0 +1,229 @@
+ADR may not necessarily work with the existing catalog CLI.
+
+## Proposal
+
+The CLI implements a set of verbs, a set of nouns, and a set of common options that may
+be used to interact with the orchestrator.
+
+The CLI follows a `verb - noun - subject` pattern. For example,
+
+- `cli get application nginx`. The verb is `get`. The noun is `application` and the subject
+  is `nginx`. This particular CLI command would return the application called nginx from the
+  Orchestrator and print them to the console in a human-readable format.
+
+### Syntax
+
+This section describes the syntax of the CLI.
+
+#### Verbs
+
+Note that some verbs may have synonyms. For example, `set` and `update` are synonyms for the same
+verb.
+
+##### Session Management
+
+- `login <user-id>` ... log in to the Orchestrator. The CLI shall prompt for a password and retrieve
+  a refresh token. The refresh token is cached locally until it is expired or a `logout` verb is used.
+
+- `logout` ... log out of the Orchestrator. Any active login state, including the refresh token, is
+  discarded.
+
+##### CRUD Operations
+
+- `list` ... lists objects on the Orchestrator. The `list` verb produces a list of all objects
+  on the Orchestrator, with an optional filter. List should return tabular summary-level information
+  about the objects.
+
+- `get` / `describe` ... retrieves objects from the Orchestrator. `get` is for retrieving single objects whereas
+  `list` is for retrieving multiple objects. `get` should retrieve more verbose information about the
+  object than `list`.
+
+- `create` ... creates an object on the Orchestrator. Additional noun-specific parameters are used to
+  describe the properties of the object to be created.
+
+- `delete` ... deletes an object from the Orchestrator.
+
+- `set` / `update` ... modifies one or more values in an existing object on the Orchestrator.
+
+- `apply` ... create or modify an object from a yaml or json specification, from stdin or from a local
+  file.
+
+##### Configuration and utility
+
+- `config` ... configures the CLI. For example, `cli config set endpoint https://my-orchestrator.com/`
+  TO-DO: This command inherited from the catalog CLI does not follow consistent syntax
+  and we should consider discarding it in favor treating configuration as any other verb-noun operation,
+  i.e. `cli update config --endpoint https://my-orchestrator.com/`.
+
+- `completion` ... return a script that can be used to configure autocompletion.
+  TO-DO: Consider treating completion as a verb/noun operation? i.e. `cli get completion bash`.
+
+- `upload <directory>` ... upload items to the Orchestrator. This is currently a catalog-specific
+  operation and uploads a directory of yaml files to the orchestrator.
+
+- `version` ... return the version number of the CLI.
+
+- `watch` ... watch an object in the Orchestrator using webhook or stream interface, returning changes
+  as they occur.
+  TO-DO: drop this?
```

Comment:
Someone could just wrap with the `watch` command, so would be in favor of dropping

Not all orchestrator interfaces have webhook/stream interfaces, and this could be expensive on the API side.

Reply:
`watch` was originally designed to exercise/test a websocket-based interface to the catalog that was going to be used to make the GUI event-driven. Let's drop it.